metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -0,0 +1,920 @@
+        appendWriteEvent("electLeaders", request.timeoutMs(),
+            () -> replicationControl.electLeaders(request));
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<FeatureManager.FinalizedFeaturesAndEpoch> finalizedFeatures() {
+        return appendReadEvent("getFinalizedFeatures",
+            () -> featureControl.finalizedFeaturesAndEpoch(lastCommittedOffset));
+    }
+
+    @Override
+    public CompletableFuture<Map<ConfigResource, ApiError>> incrementalAlterConfigs(
+        Map<ConfigResource, Map<String, Entry<OpType, String>>> configChanges,
+        boolean validateOnly) {
+        return appendWriteEvent("incrementalAlterConfigs", () -> {
+            ControllerResult<Map<ConfigResource, ApiError>> result =
+                configurationControl.incrementalAlterConfigs(configChanges);
+            if (validateOnly) {
+                return result.withoutRecords();
+            } else {
+                return result;
+            }
+        });
+    }
+
+    @Override
+    public CompletableFuture<Map<ConfigResource, ApiError>> legacyAlterConfigs(
+        Map<ConfigResource, Map<String, String>> newConfigs, boolean validateOnly) {
+        return appendWriteEvent("legacyAlterConfigs", () -> {
+            ControllerResult<Map<ConfigResource, ApiError>> result =
+                configurationControl.legacyAlterConfigs(newConfigs);
+            if (validateOnly) {
+                return result.withoutRecords();
+            } else {
+                return result;
+            }
+        });
+    }
+
+    class ProcessBrokerHeartbeatOperation
+            implements ControllerWriteOperation<BrokerHeartbeatReply> {
+        private final BrokerHeartbeatRequestData request;
+        private boolean updateShutdownOffset = false;
+
+        ProcessBrokerHeartbeatOperation(BrokerHeartbeatRequestData request) {
+            this.request = request;
+        }
+
+        @Override
+        public ControllerResult<BrokerHeartbeatReply> generateRecordsAndResult() {
+            clusterControl.checkBrokerEpoch(request.brokerId(), request.brokerEpoch());
+            List<ApiMessageAndVersion> records = new ArrayList<>();
+            boolean movingLeadersForShutDown = false;
+            if (request.wantShutDown()) {
+                records = replicationControl.removeLeaderships(request.brokerId());
+                movingLeadersForShutDown = !records.isEmpty();
+            }
+            ControllerResult<BrokerHeartbeatReply> result = clusterControl.
+                processBrokerHeartbeat(request, lastCommittedOffset,
+                    movingLeadersForShutDown);
+            if (movingLeadersForShutDown && !result.response().shouldShutDown()) {
+                updateShutdownOffset = true;
+            }
+            if (result.response().isFenced()) {
+                replicationControl.removeFromIsr(request.brokerId(), records);
```

Comment:
It seemed safer to leave it in the ISR until it's ready to shut down for good.  Also, if we take it out, it might just get re-added if it catches up... ?

Reply:
Well, in ZK based approach, in response to a controlled shutdown, the controller changes ISR and the leader. Here, it seems that `result.response().isFenced()` is not always true if the broker heartbeat indicates the intention to shut down?

Reply:
Hmm... We only fence the broker once controlled shutdown has completed.  If we fenced it immediately that would be disruptive to clients since they wouldn't be able to continue fetching from it until the leaderships have moved. Basically immediate fencing is the NON-controlled shutdown path....